polytracker/include/polytracker/tracing.h

```diff
@@ -0,0 +1,331 @@
+/*
+ * polytracker_tracing.h
+ *
+ *  Created on: Jul 3, 2020
+ *      Author: Evan Sultanik
+ */
+#ifndef POLYTRACKER_INCLUDE_POLYTRACKER_TRACING_H_
+#define POLYTRACKER_INCLUDE_POLYTRACKER_TRACING_H_
+
+#include <functional>
+#include <list>
+#include <stack>
+#include <string.h>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "dfsan/dfsan_types.h"
+#include "polytracker/basic_block_types.h"
+
+namespace polytracker {
+
+extern size_t numTraceEvents;
+
+struct TraceEvent {
+  TraceEvent *previous;
+  TraceEvent *next;
+  size_t eventIndex;
+  TraceEvent();
+  virtual ~TraceEvent() = default;
+};
+
+struct BasicBlockTrace {
+  const char *fname;
+  BBIndex index;
+  size_t entryCount;
+
+  bool operator==(const BasicBlockTrace &other) const {
+    return (fname == other.fname /* these are globals, so it should be fine
+                                  * to skip a string compare and just compare
+                                  * pointers */
+            && index == other.index && entryCount == other.entryCount);
+  }
+
+  bool operator<(const BasicBlockTrace &rhs) const {
```

Comment:
I think this might actually be dead code now, I'll have to check. This comparison not only checks that two basic blocks are the same, but also that they have been entered the same number of times in the current stack frame (the `entryCount` member).